### DISCUSSION TITLE
Components with calls to <Link>

### DISCUSSION THREAD

**Discussion by HAN20190621** [fwouts/previewjs#182] [Q&A] [answered]
Hi,
I am learning react and was looking for an extension to view my component without running the app on the browser and saw the "React Preview". I have installed it in VS Code.

I am trying to load the About in preview however I am getting an error - Error: Invariant failed you should not use <Link> outside a Router.

App invokes About  and About goes back to App. 

Does the extension "React Preview" works with components with calls to Route/Link? Thanks.

```js
import { BrowserRouter as Router, Switch, Route, Link } from 'react-router-dom';
import About from './components/About';
import Footer from './components/Footer';

const App = () => {
  return (
    <Router>
        <Switch>
          <Route path='/about' component={About} />
        </Switch>
        <Footer />
    </Router>
  );
};
```

**About component**

```js
import React from 'react';
import { Link } from 'react-router-dom';

const About = () => {
  return (
    <div style={{ textAlign: 'center' }}>
      <h4>Version 1.0.0</h4>
      <Link to='/'>Go back</Link>
    </div>
  );
};

export default About; 
```

**Comment by fwouts** [comment, marked as answer]
Hey @HAN20190621, welcome to React!

I'd encourage you to use [Preview.js](https://previewjs.com), which is going to replace React Preview very soon. They're basically the same, except Preview.js is more recent and supports frameworks other than React :)

You can definitely make this work, all you need is to set up a Wrapper file that includes the BrowserRouter component. I've added an example app so you can use that for inspiration! See https://github.com/fwouts/previewjs/tree/main/smoke-test-apps/cra-react-router

The same should also work with React Preview, the only difference being that the `__previewjs__` directory would be named `__reactpreview__` instead.

See https://previewjs.com/docs/config/wrapper for documentation.

Let me know how you go, and good luck on your learning journey!

**Comment by HAN20190621** [comment]
Thanks for this. It is loading now. I love seeing the components on the fly. I will try the new version.

If you don't mind, I have another question about useLocation().  I am using useLocation to test the path (if equal to root) however the value of location is /preview/ when viewing the component on Preview. I prefer not to change my code to also test for "/preview/". 
    
```
import { useLocation, Link } from 'react-router-dom';

const Header = ({ title, resizing }) => {
  const location = useLocation();
  return (
    <div className='header'>
      <h1>{title}</h1>
      <div>
        {location.pathname === '/' && (
          <Link to='./Start'>
            <Button text='Start' colour='green' />
          </Link>
        )}
      </div>
    </div>
  );
};

Header.propTypes = {
  title: PropTypes.string,
  resizing: PropTypes.bool,
};

Header.defaultProps = {
  title: 'Start',
  resizing: false,
};

export default Header;


```

**Reply by fwouts** [reply]
That's good to hear!

This one is a bit more complicated. I don't think there's an easy way to make this work the way you'd like, however what I can recommend is to split up your components into "dumb UI components" that are controlled by stateful components.

For example in your case, it would look like:
```js
import { useLocation, Link } from 'react-router-dom';

const DumbHeader = ({ locationPath, title, resizing }) => {
  return (
    <div className='header'>
      <h1>{title}</h1>
      <div>
        {locationPath === '/' && (
          <Link to='./Start'>
            <Button text='Start' colour='green' />
          </Link>
        )}
      </div>
    </div>
  );
};

DumbHeader.propTypes = {
  locationPath: PropTypes.string,
  title: PropTypes.string,
  resizing: PropTypes.bool,
};

const StatefulHeader = ({ title, resizing }) => {
  const location = useLocation();
  return <DumbHeader locationPath={location.pathname} title={title} resizing={resizing} />
});

StatefulHeader.defaultProps = {
  title: 'Start',
  resizing: false,
};

export default StatefulHeader;
```

This approach lends itself better to tools like Preview.js, [Viteshot](https://viteshot.com) and [Storybook](https://storybook.js.org), which are better suited for stateless components ("DumbHeader" in this case).

Note there are more sophisticated ways to organise your components—this is just one way off the top of my head.

**Comment by HAN20190621** [comment]
Thank you :)
